File: compat/suites/node-js-sdk/src/groups/shield.ts
/**
 * groups/shield.ts — Shield (DDoS protection) compatibility test groups for the Node.js suite.
 *
 * Status: NOT implemented in Overcast. All tests expected to fail with 501.
 * These tests define the coverage target for future Shield implementation.
 *
 * Groups:
 *   shield-protections — protection lifecycle
 */

import {
  CreateProtectionCommand,
  DeleteProtectionCommand,
  DescribeProtectionCommand,
  ListProtectionsCommand,
  DescribeSubscriptionCommand,
} from "@aws-sdk/client-shield";
import { makeClients } from "../lib/clients.js";
import type { TestGroup } from "../lib/harness.js";
import * as assert from "node:assert/strict";

export function makeShieldGroups(suite: string): TestGroup[] {
  return [
    // ── shield-protections ─────────────────────────────────────────────────
    {
      suite,
      service: "shield",
      name: "shield-protections",
      tests: [
        {
          name: "DescribeSubscription",
          fn: async (ctx) => {
            const { shield } = makeClients(ctx);
            // Shield Advanced requires a subscription; expect 501 in the stub.
            await shield.send(new DescribeSubscriptionCommand({}));
          },
        },
        {
          name: "CreateProtection",
          fn: async (ctx) => {
            const { shield } = makeClients(ctx);
            const resp = await shield.send(
              new CreateProtectionCommand({
                Name: `compat-${ctx.runId}`,
                ResourceArn: `arn:aws:ec2:us-east-1:000000000000:eip/eipalloc-${ctx.runId}`,
              }),
            );
            assert.ok(
              resp.ProtectionId,
              "CreateProtection: missing ProtectionId",
            );
            (ctx as Record<string, unknown>)["_protectionId"] =
              resp.ProtectionId;
          },
        },
        {
          name: "ListProtections",
          fn: async (ctx) => {
            const { shield } = makeClients(ctx);
            await shield.send(new ListProtectionsCommand({}));
          },
        },
        {
          name: "DeleteProtection",
          fn: async (ctx) => {
            const { shield } = makeClients(ctx);
            const id = (ctx as Record<string, unknown>)[
              "_protectionId"
            ] as string;
            if (!id) return;
            await shield.send(
              new DeleteProtectionCommand({ ProtectionId: id }),
            );
          },
        },
        {
          name: "DescribeProtection",
          fn: async (ctx) => {
            const { shield } = makeClients(ctx);
            // Create a fresh protection to describe (since DeleteProtection may have already deleted it).
            const createResp = await shield.send(
              new CreateProtectionCommand({
                Name: `compat-desc-${ctx.runId}`,
                ResourceArn: `arn:aws:ec2:us-east-1:000000000000:eip/eipalloc-desc-${ctx.runId}`,
              }),
            );
            const id = createResp.ProtectionId;
            assert.ok(id, "DescribeProtection: could not create protection");
            (ctx as Record<string, unknown>)["_descProtectionId"] = id;
            const resp = await shield.send(
              new DescribeProtectionCommand({ ProtectionId: id }),
            );
            assert.ok(
              resp.Protection,
              "DescribeProtection: missing Protection",
            );
            assert.strictEqual(
              resp.Protection.Id,
              id,
              `DescribeProtection: expected Id ${id}`,
            );
          },
        },
      ],
      teardown: async (ctx) => {
        const { shield } = makeClients(ctx);
        const id = (ctx as Record<string, unknown>)["_protectionId"] as string;
        if (id) {
          try {
            await shield.send(
              new DeleteProtectionCommand({ ProtectionId: id }),
            );
          } catch {}
        }
        const descId = (ctx as Record<string, unknown>)[
          "_descProtectionId"
        ] as string;
        if (descId) {
          try {
            await shield.send(
              new DeleteProtectionCommand({ ProtectionId: descId }),
            );
          } catch {}
        }
      },
    },
  ];
}

File: compat/suites/python-sdk/groups/shield.py
"""
groups/shield.py — Shield (DDoS protection) compatibility test implementations for the Python suite.
"""

from __future__ import annotations
from lib.harness import TestContext
from lib.clients import make_clients


def _shield(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region)._get("shield")


# ── shield-protections ────────────────────────────────────────────────────────

def DescribeSubscription(ctx: TestContext) -> None:
    _shield(ctx).describe_subscription()


def CreateProtection(ctx: TestContext) -> None:
    sh = _shield(ctx)
    name = f"compat-{ctx.run_id}"
    resource_arn = f"arn:aws:ec2:us-east-1:000000000000:eip/eipalloc-{ctx.run_id}"
    resp = sh.create_protection(Name=name, ResourceArn=resource_arn)
    if not resp.get("ProtectionId"):
        raise AssertionError("CreateProtection: missing ProtectionId")
    ctx["shield_protection_id"] = resp["ProtectionId"]


def ListProtections(ctx: TestContext) -> None:
    _shield(ctx).list_protections()


def DescribeProtection(ctx: TestContext) -> None:
    sh = _shield(ctx)
    protection_id = ctx.get("shield_protection_id")
    if not protection_id:
        raise AssertionError("DescribeProtection: no protection_id")
    resp = sh.describe_protection(ProtectionId=protection_id)
    prot = resp.get("Protection", {})
    if prot.get("Id") != protection_id:
        raise AssertionError(f"DescribeProtection: id mismatch")


def DeleteProtection(ctx: TestContext) -> None:
    sh = _shield(ctx)
    protection_id = ctx.get("shield_protection_id")
    if not protection_id:
        return
    sh.delete_protection(ProtectionId=protection_id)


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "DescribeSubscription": DescribeSubscription,
    "CreateProtection": CreateProtection,
    "ListProtections": ListProtections,
    "DescribeProtection": DescribeProtection,
    "DeleteProtection": DeleteProtection,
}

SETUP = {}
TEARDOWN = {
    "shield-protections": lambda ctx: _teardown_protection(ctx),
}


def _teardown_protection(ctx: TestContext) -> None:
    protection_id = ctx.get("shield_protection_id")
    if protection_id:
        try:
            _shield(ctx).delete_protection(ProtectionId=protection_id)
        except Exception:
            pass
